File: lib/rgfa/line/group/gfa2/references.rb
RGFA::Line::Group::GFA2 ||= Module.new

module RGFA::Line::Group::GFA2::References

  private

  def prepare_and_check_ref(ref)
    ref = line_for_ref_symbol(ref) if ref.kind_of?(Symbol)
    check_ref_class(ref)
    check_ref_connection(ref)
    check_ref_not_self(ref)
    return ref
  end

  def check_ref_class(item)
    if ![RGFA::Line::Edge::GFA2,
         RGFA::Line::Segment::GFA2,
         RGFA::Line::Gap,
         RGFA::Line::Group::Ordered,
         self::class].include?(item.class)
    raise RGFA::ArgumentError,
      "Line: #{self}\n"+
    "Cannot add items of class #{item.class}\n"+
    "Only GFA2 edges, segments, gaps, groups[*] "+
    "can be added\n(* = unordered groups to unordered groups only)."
    end
  end

  def check_ref_connection(item)
    if line.rgfa != self.rgfa
      raise RGFA::ArgumentError,
        "Line: #{self}\n"+
      "Item: #{item.inspect}"+
      "The item added to the group must be connected\n"+
      "to the same RGFA object as the group"
    end
  end

  def check_ref_not_self(item)
    if (line == self)
      raise RGFA::RuntimeError,
        "Line: #{self}\n"+
      "Item is the line itself\n"+
      "A group is not allowed to refer to itself"
    end
  end

  def line_for_ref_symbol(ref)
    line = @rgfa.search_by_name(ref)
    if line.nil?
      if @rgfa.segments_first_order
        raise RGFA::NotFoundError, "Group: #{self}\n"+
        "requires a non-existing ref with ID #{ref}"
      end
      line = RGFA::Line::Unknown.new({:name => ref}, virtual: true,
                                     version: :gfa2)
      @rgfa << line
    end
    line.add_reference(self, (record_type == :O) ?
                         :ordered_groups : :unordered_groups)
    return line
  end

end

File: gfapy/line/group/gfa2/references.py
import gfapy

class References:

  def _prepare_and_check_ref(self, ref):
    if isinstance(ref, str):
      ref = self._line_for_ref_symbol(ref)
    self._check_ref_class(ref)
    self._check_ref_connection(ref)
    self._check_ref_not_self(ref)
    return ref

  def _check_ref_class(self, item):
    if item.__class__ not in [
         gfapy.line.edge.GFA2,
         gfapy.line.segment.GFA2,
         gfapy.line.gap.Gap,
         gfapy.line.group.Ordered,
         self.__class__]:
      raise gfapy.ArgumentError(
        "Line: {}\n".format(self)+
        "Cannot add items of class {}\n".format(item.__class__.__name__)+
        "Only GFA2 edges, segments, gaps, groups[*] "+
        "can be added\n(* = unordered groups to unordered groups only).")

  def _check_ref_connection(self, item):
    if item.line.gfa != self._gfa:
      raise gfapy.ArgumentError(
        "Line: {}\n".format(self)+
        "Item: {}".format(repr(item))+
        "The item added to the group must be connected\n"+
        "to the same GFA object as the group")

  def _check_ref_not_self(self, item):
    if (item.line == self):
      raise gfapy.RuntimeError(
        "Line: {}\n".format(self)+
        "Item is the line itself\n"+
        "A group is not allowed to refer to itself")

  def _line_for_ref_symbol(self, ref):
    line = self._gfa.line(ref)
    if line is None:
      if self._gfa._segments_first_order:
        raise gfapy.NotFoundError("Group: {}\n".format(self)+
                        "requires a non-existing ref with ID {}".format(ref))
      line = gfapy.line.unknown.Unknown({"name" : ref}, virtual = True,
                                        version = "gfa2")
      self._gfa.add_line(line)
    line._add_reference(self, "paths" if (self.record_type == "O") else "sets")
    return line
